contracts/v2/SwapExchange.sol

```diff
@@ -0,0 +1,145 @@
+        external returns (uint[] memory amounts);
+    function swapExactTokensForETH(uint amountIn, uint amountOutMin, address[] calldata path, address to, uint deadline)
+        external returns (uint[] memory amounts);
+
+    // Token1 => Token2
+    function swapTokensForExactTokens(uint amountOut, uint amountInMax, address[] calldata path, address to, uint deadline)
+        external returns (uint[] memory amounts);
+    function swapExactTokensForTokens(uint amountIn, uint amountOutMin, address[] calldata path, address to, uint deadline)
+        external returns (uint[] memory amounts);
+
+    function WETH() external view returns (address);
+
+    function getAmountsOut(uint amountIn, address[] calldata path) external view returns (uint[] memory amounts);
+    function getAmountsIn(uint amountOut, address[] calldata path) external view returns (uint[] memory amounts);
+}
+
+/**
+ * @dev SwapExchage is used to deposit/withdraw App Coins based on native tokens.
+ */
+contract SwapExchage {
+    AppCoin private appCoin;
+    ISwap private swap;
+
+    modifier onlyInitialized() {
+        require(address(appCoin) != address(0), "SwapExchage: uninitialized yet");
+        _;
+    }
+
+    function initialize(AppCoin appCoin_, ISwap swap_) public {
+        require(address(appCoin) == address(0), "SwapExchage: already initialized");
+
+        appCoin = appCoin_;
+        swap = swap_;
+    }
+
+    /**
+     * @dev Preview how many ETH required to deposit `amount` of App Coins.
+     *
+     * Parameters:
+     * - amount: amount of App Coins to deposit.
+     */
+    function previewDepositETH(uint256 amount) public view onlyInitialized returns (uint256) {
+        amount = appCoin.previewMint(amount);
+
+        address[] memory path = new address[](2);
+        path[0] = swap.WETH();
+        path[1] = appCoin.asset();
+
+        uint[] memory amounts = swap.getAmountsIn(amount, path);
+
+        return amounts[0];
+    }
+
+    /**
+     * @dev Deposit `amount` of App Coins to `receiver` with ETH.
+     *
+     * Parameters:
+     * - amount: amount of App Coins to deposit.
+     * - receiver: address to receive the App Coins.
+     */
+    function depositETH(uint256 amount, address receiver) public payable onlyInitialized {
+        amount = appCoin.previewMint(amount);
+
+        uint256 balanceBefore = address(this).balance - msg.value;
+
+        address[] memory path = new address[](2);
+        path[0] = swap.WETH();
+        path[1] = appCoin.asset();
+
+        // swap ETH for tokens
+        swap.swapETHForExactTokens{value: msg.value}(amount, path, receiver, block.timestamp);
+
+        // approve and deposit for receiver
+        SafeERC20.safeApprove(IERC20(appCoin.asset()), address(appCoin), amount);
+        appCoin.deposit(amount, receiver);
+
+        // refund dust ETH if any
+        uint256 dust = address(this).balance - balanceBefore;
```

Comment:
Check consumed amount of eth in return value from swapETHForExactTokens

Reply:
Relying on contract balance is more error tolerable, we can have more discussion offline.

Reply:
In addition, add `ReentrancyGuard` to prevent malicious callback from `receiver` contract.